net: lwm2m: Allow cancel-observe to match path

endian-benjamin opened the pull request:
The specification states that the server can cancel observations "at any
moment, by sendinga GET request with Observe option=1, the LwM2M Server
cancancel an “Observe”operation on a specified Resource, or specified
Object Instance(s)."

It does not mention any token matching requirement. The EMQx LwM2M
implementation uses a new token for instance, which does not
work with Zephyrs token matching cancel-observe.

This commit introduces cancel-observe via path matching. This could
hypothetically introduce problems when we are connected to multiple
peers simultaneously, but since that is not likely to be supported for a
long time (if ever), this change should be fairly uncontroversial since
path matching is only used as a fallback.

endian-benjamin committed:
net: lwm2m: Allow cancel-observe to match path

The specification states that the server can cancel observations "at any
moment, by sendinga GET request with Observe option=1, the LwM2M Server
cancancel an “Observe”operation on a specified Resource, or specified
Object Instance(s)."

It does not mention any token matching requirement, but RFC 7641 does.
The correct interpretation is not obvious. The EMQx LwM2M implementation
uses a new token for instance, which does not work with Zephyrs token
matching cancel-observe.

This commit introduces cancel-observe via path matching as a Kconfig
option. This could hypothetically introduce problems when we are
connected to multiple peers simultaneously, but since that is not likely
to be supported for a long time (if ever), this change should be fairly
uncontroversial since path matching is only used as a fallback.

Signed-off-by: Benjamin Lindqvist <benjamin.lindqvist@endian.se>

rlubos commented:
> It does not mention any token matching requirement. The EMQx LwM2M
> implementation uses a new token for instance, which does not
> work with Zephyrs token matching cancel-observe.

This is covered by the RFC 7641, which LwM2M is based on:

>In some circumstances, it may be desirable to cancel an observation
   and release the resources allocated by the server to it more eagerly.
   In this case, a client MAY explicitly deregister by issuing a GET
   request that has the **Token field set to the token of the observation
   to be cancelled** and includes an Observe Option with the value set to
   1 (deregister).

So it's not something Zephyr specific to match the token, but it's required by the standard. I'm not familiar with the EMQx LwM2M implementation, but I guess it should be rather reported on their side that they do not follow the spec.

> This commit introduces cancel-observe via path matching. This could
> hypothetically introduce problems when we are connected to multiple
> peers simultaneously, but since that is not likely to be supported for a
> long time (if ever), this change should be fairly uncontroversial since
> path matching is only used as a fallback.

I'd rather cautious with assumptions like this and would rather avoid putting potential obstacles for future imporvements of the LwM2M stack. But on the other hand, in the mutli-server scenario the observe node would likely need to store the information about the observation owner so it could be verified in such case to avoid a situation where one server cancels others observation. Still, as a non-standard extension I'd consider making this configurable with Kconfig.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -575,6 +575,34 @@ static int engine_remove_observer(const uint8_t *token, uint8_t tkl)
+		return -ENOENT;
+	}
+
+	LOG_INF("Removing observer for path %u/%u/%u/%u",
```
This would suggest that Resource instance was always cancelled (which isn't even a valid observation), at least the path level should be included in the log, or even better some smart formatting applied (to print the actual path cancelled).

endian-benjamin replied:
sure, good feedback

endian-benjamin commented:
> So it's not something Zephyr specific to match the token, but it's required by the standard. I'm not familiar with the EMQx LwM2M implementation, but I guess it should be rather reported on their side that they do not follow the spec.

I definitely agree that the EMQx implementation SHOULD cache the token and cancel using that, but I'm not 100% sure I agree with your interpretation of the specification. LwM2M has a separate mechanism for uniquely identifying an observe (that is, the resource path). Since that is not a part of the original CoAP spec, it is natural that token matching is required, *for CoAP*. Path-based cancel-observe could be an intentional relaxation on OMA:s part.

> I'd rather cautious with assumptions like this and would rather avoid putting potential obstacles for future imporvements of the LwM2M stack. But on the other hand, in the mutli-server scenario the observe node would likely need to store the information about the observation owner so it could be verified in such case to avoid a situation where one server cancels others observation. Still, as a non-standard extension I'd consider making this configurable with Kconfig.

...but in any case, you are right of course - a Kconfig is strictly better.

rlubos commented:
> I'm not 100% sure I agree with your interpretation of the specification. LwM2M has a separate mechanism for uniquely identifying an observe (that is, the resource path). Since that is not a part of the original CoAP spec, it is natural that token matching is required, _for CoAP_. Path-based cancel-observe could be an intentional relaxation on OMA:s part.

Sorry but I disagree, I think there's no ambiguity here. I haven't found a single reference in the LwM2M spec that suggests that LwM2M implements some non-standard Observe mechanism. Instead, we find in there:
> Periodic and event-triggered reporting about Resource values from the LwM2M Client to the LwM2M Server is achieved
through CoAP Observation [OBSERVE].

Where `[OBSERVE]` is a reference to the RFC 7641. There is nothing unique to LwM2M in specifying the resource path, the aforementioned RFC states (chapter 3.1):
> A client registers its interest in a **resource** by issuing a GET
   request with an Observe Option set to 0 (register).

Even if you look at the examples in the RFC, the resource URI is always specified in the Observe request. The only confusion I can see is that from CoAP perspective a resource is any data object that can be identified by an URI, so a LwM2M Object or Object Instance would also be interpreted as CoAP resource. But otherwise, it's pretty clear to me that in LwM2M we deal with the standard Observe mechanism. 

Of course, I could have overlooked something in the LwM2M spec or RFCs so fell free to correct me if you disagree.

endian-benjamin commented:
> Even if you look at the examples in the RFC, the resource URI is always specified in the Observe request. The only confusion I can see is that from CoAP perspective a resource is any data object that can be identified by an URI, so a LwM2M Object or Object Instance would also be interpreted as CoAP resource. But otherwise, it's pretty clear to me that in LwM2M we deal with the standard Observe mechanism.

Well it's hard to find good counterarguments here, it's just that the actual language of the LwM2M spec makes it seem as if the intention is to allow path matching as an ALTERNATIVE - I mean why introduce path matching at all if it's not meant as an alternative to token matching, right? It just makes no intuitive sense.

Anyways, I rewrote is as Kconfigurable and added a fancy path macro. I need to test it a bit tho.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -575,6 +575,54 @@ static int engine_remove_observer(const uint8_t *token, uint8_t tkl)
+		return -ENOENT;
+	}
+
+	LOG_INF("Removing observer for path %s", path);
```
`lwm2m_path_log_strdup(path)`?

endian-benjamin replied:
Yeah I fixed it but force pushed to the wrong branch :D

rlubos reviewed (commented):
Just one issue spotted.

rlubos commented:
Compliance check also spotted a few issues, please have a look.

endian-benjamin commented:
Hopefully ok now

endian-benjamin commented:
> > Even if you look at the examples in the RFC, the resource URI is always specified in the Observe request. The only confusion I can see is that from CoAP perspective a resource is any data object that can be identified by an URI, so a LwM2M Object or Object Instance would also be interpreted as CoAP resource. But otherwise, it's pretty clear to me that in LwM2M we deal with the standard Observe mechanism.
> 
> Well it's hard to find good counterarguments here, it's just that the actual language of the LwM2M spec makes it seem as if the intention is to allow path matching as an ALTERNATIVE - I mean why introduce path matching at all if it's not meant as an alternative to token matching, right? It just makes no intuitive sense.
> 
> Anyways, I rewrote is as Kconfigurable and added a fancy path macro. I need to test it a bit tho.

A colleague had a good point - apparently

endian-benjamin commented:
misclick close haha sorry

rlubos reviewed (approved):
LGTM. The CI faliure seems to be a random one.

endian-benjamin commented:
What's going on here, can we retrigger the random ci failure, or merge anyway?

rlubos commented:
> What's going on here, can we retrigger the random ci failure, or merge anyway?

You can force-push the branch or close/repoen the PR to retrigger the CI.

jukkar commented:
It is a bit weird why the unit test fails. @endian-benjamin could you rebase against latest upstream and resubmit, perhaps it solves the issue.

jukkar commented:
ping @endian-benjamin

endian-benjamin commented:
Thanks for the ping, didn't realize it was broken
